pkg/pillar/cmd/zedrouter/networkinstance.go

```diff
@@ -1870,3 +1889,273 @@ func uplinkToPhysdev(ifname string) string {
+	if status.HostIPAddrModInitialized {
+		return
+	}
+	log.Functionf("initializeIPAddrMgmtModule(%s-%s)\n",
+		status.DisplayName, status.UUID.String())
+	if err := doNetworkInstanceStatusDhcpRangeSanityCheck(status); err == nil {
+		log.Functionf("Processing DhcpRange (%s - %s)\n",
+			status.DhcpRange.Start.String(), status.DhcpRange.End.String())
+		setHostIPAddrCountOnDhcpRange(status)
+	} else {
+		if err := doNetworkInstanceSubnetSanityCheck(ctx, status); err == nil {
+			log.Functionf("Processing Subnet (%s/%s)\n",
+				status.Subnet.IP.String(), status.Subnet.Mask.String())
+			setHostIPAddrCountOnSubnet(status)
+		}
+	}
+	// get mappings from persistent to Bitmap across reboot,
+	// deactivate/activate cycle
+	persistentAddrMapOnNetworkInstanceRestore(ctx, status)
+	status.HostIPAddrModInitialized = true
+	publishNetworkInstanceStatus(ctx, status)
+}
+
+// Assignable IP Address counter routines
+// for DhcpRange
+func setHostIPAddrCountOnDhcpRange(status *types.NetworkInstanceStatus) {
+	addr1 := status.DhcpRange.Start.To4()
+	addr2 := status.DhcpRange.End.To4()
+	if addr1 == nil || addr2 == nil {
+		return
+	}
+	start := uint(addr1[0])<<24 + uint(addr1[1])<<16 +
+		uint(addr1[2])<<8 + uint(addr1[3])
+	end := uint(addr2[0])<<24 + uint(addr2[1])<<16 +
+		uint(addr2[2])<<8 + uint(addr2[3])
+	addrcount := int(end + 1 - start)
+	status.HostIPAddressCount = getMaxIPAddrCountForNetworkInstance(addrcount)
+}
+
+// for Subnet
+func setHostIPAddrCountOnSubnet(status *types.NetworkInstanceStatus) {
+	maskLen, _ := status.Subnet.Mask.Size()
+	addrcount := 1 << (32 - maskLen)
+	// especially applicable for smaller subnets,
+	// discount the network and broadcast addresses(2)
+	// for larger subnets, it will be restricted to 254
+	// by getMaxIPAddrCountForNetworkInstance
+	addrcount -= 2
+	status.HostIPAddressCount = getMaxIPAddrCountForNetworkInstance(addrcount)
+}
+
+// for network size more than types.AppNumMax(254), restrict to
+// types.AppNumMax(254), limited by bitmap size, with a discount of 2
+// (network address and broadcast address)
+func getMaxIPAddrCountForNetworkInstance(count int) int {
+	if count > types.AppNumMax {
+		return types.AppNumMax
+	}
+	return count
+}
+
+// static ip address to app number
+func uNetAppNumForStaticGet(ctx *zedrouterContext,
+	status *types.NetworkInstanceStatus, appID string,
+	staticIPAddr net.IP) (int, error) {
+	if status.HostIPAddressCount == 0 {
+		return types.AppNumInvalid, nil
+	}
+	appNumOnUNet, err := persistentAddrMapAppNumGet(ctx, status.UUID.String(), appID)
+	if err == nil {
+		return appNumOnUNet, nil
+	}
+	startIP := status.DhcpRange.Start.To4()
+	if startIP == nil {
+		return types.AppNumInvalid, errors.New("parse error")
+	}
+	start := uint(startIP[0])<<24 + uint(startIP[1])<<16 +
+		uint(startIP[2])<<8 + uint(startIP[3])
+	ipAddr := staticIPAddr.To4()
+	addr := uint(ipAddr[0])<<24 + uint(ipAddr[1])<<16 +
+		uint(ipAddr[2])<<8 + uint(ipAddr[3])
+	appNumOnUNet = int(addr - start)
```

Comment:
This is an interesting approach to handle static, but you can end up with some arbitrary number including a negative number.